server/datastore/mysql/queries.go

```diff
@@ -283,62 +311,93 @@ func (ds *Datastore) NewQuery(
 		return ctxerr.New(ctx, "updateQueryLabelsInTx called with nil tx")
 	}
+	if len(queries) == 0 {
+		return nil
+	}
 
-	var err error
-
-	insertLabelSql := `
-		INSERT INTO query_labels (
-			query_id,
-			label_id
-		)
-		SELECT ?, id
-		FROM labels
-		WHERE name IN (?)
-	`
-
-	deleteLabelStmt := `
-		DELETE FROM query_labels
-		WHERE query_id = ?
-	`
+	queriesIDs := make([]uint, 0, len(queries))
+	for _, q := range queries {
+		queriesIDs = append(queriesIDs, q.ID)
+	}
 
-	_, err = tx.ExecContext(ctx, deleteLabelStmt, query.ID)
+	deleteQueryLabelsStm, args, err := sqlx.In(`DELETE FROM query_labels WHERE query_id IN (?)`, queriesIDs)
 	if err != nil {
-		return ctxerr.Wrap(ctx, err, "removing old query labels")
+		return ctxerr.Wrap(ctx, err, "deleting old query labels")
+	}
+	if _, err := tx.ExecContext(ctx, deleteQueryLabelsStm, args...); err != nil {
+		return ctxerr.Wrap(ctx, err, "deleting old query labels")
 	}
 
-	if len(query.LabelsIncludeAny) == 0 {
+	var lblNames []interface{}
+	for _, q := range queries {
+		for _, lbl := range q.LabelsIncludeAny {
+			lblNames = append(lblNames, lbl.LabelName)
+		}
+	}
+	if len(lblNames) == 0 {
 		return nil
 	}
 
-	labelNames := []string{}
-	for _, label := range query.LabelsIncludeAny {
-		labelNames = append(labelNames, label.LabelName)
+	// We need to figure out the label IDs for the labels we're going to add.
+	stm, args, err := sqlx.In(`SELECT id, name FROM labels WHERE name IN (?)`, lblNames)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "fetching label IDs")
 	}
 
-	labelStmt, args, err := sqlx.In(insertLabelSql, query.ID, labelNames)
+	rows, err := tx.QueryxContext(ctx, stm, args...)
 	if err != nil {
-		return ctxerr.Wrap(ctx, err, "creating query label update statement")
+		return ctxerr.Wrap(ctx, err, "fetching label IDs")
+	}
+	defer rows.Close()
+
+	lblNameToID := make(map[string]uint)
+	for rows.Next() {
+		var id uint
+		var name string
+		if err := rows.Scan(&id, &name); err != nil {
+			return ctxerr.Wrap(ctx, err, "scan existing query")
+		}
+		lblNameToID[name] = id
+	}
+	if err := rows.Err(); err != nil {
+		return ctxerr.Wrap(ctx, err, "fetching query IDs")
+	}
+
+	params := make([]string, 0, len(lblNames))
+	args = make([]interface{}, 0, len(lblNames)*2)
+	for _, q := range queries {
+		for _, lbl := range q.LabelsIncludeAny {
+			if lblID, ok := lblNameToID[lbl.LabelName]; ok {
+				params = append(params, "(?, ?)")
+				args = append(args, q.ID, lblID)
+			}
+		}
+	}
+
+	if len(params) == 0 {
+		return nil
 	}
 
-	if _, err := tx.ExecContext(ctx, labelStmt, args...); err != nil {
+	insertSQL := fmt.Sprintf(`INSERT INTO query_labels (query_id, label_id) VALUES %s`, strings.Join(params, ", "))
+	if _, err := tx.ExecContext(ctx, insertSQL, args...); err != nil {
 		return ctxerr.Wrap(ctx, err, "creating query labels")
 	}
 
-	if err := loadLabelsForQueries(ctx, tx, []*fleet.Query{query}); err != nil {
+	if err := loadLabelsForQueries(ctx, tx, queries); err != nil {
```

Comment:
Do you need to reload this information from the DB? (Don't you have query ID and label ID in memory already?)